core/rawdb/accessors_chain.go

```diff
@@ -485,24 +482,104 @@ func ReadReceipts(db DatabaseReader, hash common.Hash, number uint64) types.Rece
-	newV := make([]byte, 0, 1024)
-	err := cbor.Marshal(&newV, receipts)
+func WriteReceipts(tx DatabaseWriter, number uint64, receipts types.Receipts) error {
+	buf := bytes.NewBuffer(make([]byte, 0, 1024))
+	for txId, r := range receipts {
+		if len(r.Logs) == 0 {
+			continue
+		}
+
+		buf.Reset()
+		err := cbor.Marshal(buf, r.Logs)
+		if err != nil {
+			return fmt.Errorf("encode block logs for block %d: %v", number, err)
+		}
+
+		if err = tx.Put(dbutils.Log, dbutils.LogKey(number, uint32(txId)), buf.Bytes()); err != nil {
+			return fmt.Errorf("writing logs for block %d: %v", number, err)
+		}
+	}
+
+	buf.Reset()
+	err := cbor.Marshal(buf, receipts)
 	if err != nil {
-		log.Crit("Failed to encode block receipts", "err", err)
+		return fmt.Errorf("encode block receipts for block %d: %v", number, err)
 	}
 
-	// Store the flattened receipt slice
-	if err := db.Put(dbutils.BlockReceiptsPrefix, dbutils.BlockReceiptsKey(number, hash), newV); err != nil {
-		log.Crit("Failed to store block receipts", "err", err)
+	if err = tx.Put(dbutils.BlockReceiptsPrefix, dbutils.ReceiptsKey(number), buf.Bytes()); err != nil {
+		return fmt.Errorf("writing receipts for block %d: %v", number, err)
 	}
+	return nil
+}
+
+// WriteReceipts stores all the transaction receipts belonging to a block.
+func AppendReceipts(tx ethdb.DbWithPendingMutations, blockNumber uint64, receipts types.Receipts) error {
+	buf := bytes.NewBuffer(make([]byte, 0, 1024))
+	for txId, r := range receipts {
+		if len(r.Logs) == 0 {
+			continue
+		}
+
+		buf.Reset()
+		err := cbor.Marshal(buf, r.Logs)
+		if err != nil {
+			return fmt.Errorf("encode block receipts for block %d: %v", blockNumber, err)
+		}
+
+		if err = tx.Append(dbutils.Log, dbutils.LogKey(blockNumber, uint32(txId)), buf.Bytes()); err != nil {
+			return fmt.Errorf("writing receipts for block %d: %v", blockNumber, err)
+		}
+	}
+
+	buf.Reset()
+	err := cbor.Marshal(buf, receipts)
+	if err != nil {
+		return fmt.Errorf("encode block receipts for block %d: %v", blockNumber, err)
+	}
+
+	if err = tx.Append(dbutils.BlockReceiptsPrefix, dbutils.ReceiptsKey(blockNumber), buf.Bytes()); err != nil {
+		return fmt.Errorf("writing receipts for block %d: %v", blockNumber, err)
+	}
+	return nil
 }
 
 // DeleteReceipts removes all receipt data associated with a block hash.
-func DeleteReceipts(db DatabaseDeleter, hash common.Hash, number uint64) {
-	if err := db.Delete(dbutils.BlockReceiptsPrefix, dbutils.BlockReceiptsKey(number, hash)); err != nil {
-		log.Crit("Failed to delete block receipts", "err", err)
+func DeleteReceipts(db ethdb.Database, number uint64) error {
+	if err := db.Delete(dbutils.BlockReceiptsPrefix, dbutils.ReceiptsKey(number)); err != nil {
+		return fmt.Errorf("receipts delete failed: %d, %w", number, err)
 	}
+
+	if err := db.Walk(dbutils.Log, dbutils.LogKey(number, 0), 8*8, func(k, v []byte) (bool, error) {
+		if err := db.Delete(dbutils.Log, k); err != nil {
```

Comment:
Is deleting from the Walk safe?

Reply:
No. Better avoid - Because TxDb using same cursor. 
But also maybe make sense if TxDb will create/close new cursor inside walk - because now even .Get inside walk will change position of walk cursor.
Will fix today.